[CodeCompletion] Cache 'async'-ness of free functions

ahoppen opened the pull request:
Store whether a result is async in the `ContextFreeCodeCompletionResult` and determine whether an async method is used in a sync context when promoting the context free result to a contextual result.

rdar://78317170

ahoppen commented:
@swift-ci Please smoke test

SerenaKit commented:
Curious, what is a 'free function'?

SerenaKit commented:
I see, interesting, and why not cache the `async` status of member functions aswell?

ahoppen commented:
Because the current code completion cache is only for global (aka. top-level) results and since we don’t cache member functions we don’t need to cache their async-ness either 😉

rintaro commented on lib/IDE/CodeCompletionDiagnostics.cpp:
```diff
@@ -157,31 +157,3 @@ bool swift::ide::getContextFreeCompletionDiagnostics(
   return true;
 }
-
-bool swift::ide::getContextualCompletionDiagnostics(
```
I'm not a fan of this mismatched implementation between context-fee and contextual.
What do we want to avoid here? Couldn't `getContextualCompletionDiagnostics()` receive ASTContext and NameForDiagnostics?

rintaro commented on lib/IDE/CodeCompletionDiagnostics.cpp:
```diff
@@ -157,31 +157,3 @@ bool swift::ide::getContextFreeCompletionDiagnostics(
-  case ContextualNotRecommendedReason::InvalidAsyncContext:
-    // FIXME: Could we use 'diag::async_in_nonasync_function'?
-    return Diag.getDiagnostics(
-        Severity, Out, diag::ide_async_in_nonasync_context, D->getName());
```
_If_ we decide not to use DiagnosticEngine, remove these unused `diag::` definitions.

rintaro commented on lib/IDE/CodeCompletionResult.cpp:
```diff
@@ -71,18 +75,31 @@ ContextFreeCodeCompletionResult::createLiteralResult(
+static NullTerminatedStringRef
+getDeclNameForDiagnostics(const Decl *D, CodeCompletionResultSink &Sink) {
+  if (auto VD = dyn_cast<ValueDecl>(D)) {
+    llvm::SmallString<64> Scratch;
+    return NullTerminatedStringRef(VD->getName().getString(Scratch),
+                                   Sink.getAllocator());
```
An implementation equivalent to the current implementation would be:
```suggestion
    llvm::SmallString<64> Scratch;
    llvm::raw_svector_ostream OS(Scratch);
    VD->getName().printPretty(OS);
    return NullTerminatedStringRef(Scratch, Sink.getAllocator());
```

rintaro commented:
Sorry for the late review!

ahoppen commented on lib/IDE/CompletionLookup.cpp:
```diff
@@ -824,14 +824,12 @@ void CompletionLookup::addVarDeclRef(const VarDecl *VD,
   CodeCompletionResultBuilder Builder(
       Sink, CodeCompletionResultKind::Declaration,
       getSemanticContext(VD, Reason, dynamicLookupInfo));
+  Builder.setIsAsync(implicitlyAsync);
```
Good catch. Thanks.

ahoppen committed:
[CodeCompletion] Lazily compute contextual diagnostics

ahoppen committed:
[CodeCompletion] Compute InvalidAsyncContext warning when making a completion result contextual

Store whether a result is async in the `ContextFreeCodeCompletionResult` and determine whether an async method is used in a sync context when promoting the context free result to a contextual result.

rdar://78317170

ahoppen committed:
[CodeComplete] Store the compiler instace as a shared pointer in CompletionInstance

ahoppen commented:
@swift-ci Please smoke test
